<a href="https://colab.research.google.com/github/ACoolA-008/Data_Science_Projects/blob/main/Fake_News_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1: Fake news

0: Real news

Import dependencies

In [3]:
import numpy as np #making arrays
import pandas as pd #making data frames
import re #regular expression
from nltk.corpus import stopwords #natural language toolkit
from nltk.stem.porter import PorterStemmer 
from sklearn.feature_extraction.text import TfidfVectorizer #turn text into feature vectors
from sklearn.model_selection import train_test_split #split training and test data
from sklearn.linear_model import LogisticRegression 
from sklearn.metrics import accuracy_score

In [4]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [5]:
# stopwords
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

Data Pre-Processing

In [32]:
# Loading the dataset to pandas df
news_dataset = pd.read_csv('/content/train.csv')

In [8]:
news_dataset.shape

(20800, 5)

In [9]:
news_dataset.head()

id  ... label
0   0  ...     1
1   1  ...     0
2   2  ...     1
3   3  ...     1
4   4  ...     1

[5 rows x 5 columns]

In [11]:
# counting the number of missing values in the news_dataset
news_dataset.isnull().sum()

id           0
title      558
author    1957
text        39
label        0
dtype: int64

In [33]:
# Replacting the null values with empty strings
news_dataset = news_dataset.fillna('')

In [34]:
# merging columns of author names and news titles
news_dataset['content'] = news_dataset['author'] + ' ' + news_dataset['title']

In [14]:
print(news_dataset['content'])

0        Darrell Lucus House Dem Aide: We Didn’t Even S...
1        Daniel J. Flynn FLYNN: Hillary Clinton, Big Wo...
2        Consortiumnews.com Why the Truth Might Get You...
3        Jessica Purkiss 15 Civilians Killed In Single ...
4        Howard Portnoy Iranian woman jailed for fictio...
                               ...                        
20795    Jerome Hudson Rapper T.I.: Trump a ’Poster Chi...
20796    Benjamin Hoffman N.F.L. Playoffs: Schedule, Ma...
20797    Michael J. de la Merced and Rachel Abrams Macy...
20798    Alex Ansary NATO, Russia To Hold Parallel Exer...
20799              David Swanson What Keeps the F-35 Alive
Name: content, Length: 20800, dtype: object


In [35]:
# separting label and the rest of the data
# remove row: axis=0
# remove column: axis=1
X = news_dataset.drop(columns='label', axis=1)
Y = news_dataset['label']


Stemming:

Process of reducing a word to its root word

eg. actor, actress, acting --> act

In [36]:
port_stem = PorterStemmer()

Creating a helper function that 
1. removes all the stopwords and 
2. retrieve all the stem words


In [37]:
def stemming(content):
  stemmed_content = re.sub('[^a-zA-Z]',' ', content)
  stemmed_content = stemmed_content.lower()
  stemmed_content = stemmed_content.split()
  stemmed_content = [port_stem.stem(word) for word in stemmed_content if not word in stopwords.words('english')]
  stemmed_content = ' '.join(stemmed_content)
  return stemmed_content

Calling this helper method 

In [38]:
news_dataset['content'] = news_dataset['content'].apply(stemming)

In [39]:
print(news_dataset['content'])

0        darrel lucu hous dem aid even see comey letter...
1        daniel j flynn flynn hillari clinton big woman...
2                   consortiumnew com truth might get fire
3        jessica purkiss civilian kill singl us airstri...
4        howard portnoy iranian woman jail fiction unpu...
                               ...                        
20795    jerom hudson rapper trump poster child white s...
20796    benjamin hoffman n f l playoff schedul matchup...
20797    michael j de la merc rachel abram maci said re...
20798    alex ansari nato russia hold parallel exercis ...
20799                            david swanson keep f aliv
Name: content, Length: 20800, dtype: object


In [45]:
# Separating the data and the label
X = news_dataset['content'].values
Y= news_dataset['label'].values

In [46]:
print(X)

['darrel lucu hous dem aid even see comey letter jason chaffetz tweet'
 'daniel j flynn flynn hillari clinton big woman campu breitbart'
 'consortiumnew com truth might get fire' ...
 'michael j de la merc rachel abram maci said receiv takeov approach hudson bay new york time'
 'alex ansari nato russia hold parallel exercis balkan'
 'david swanson keep f aliv']


In [47]:
print(Y)

[1 0 1 ... 0 1 1]


In [48]:
# Converting text data to numerical data
vectorizer = TfidfVectorizer()
vectorizer.fit(X)

X = vectorizer.transform(X)

Splitting news_dataset to training and test data

In [49]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2, stratify=Y, random_state=2)

Training the model: Logistic Regression(for binary classification)

In [50]:
model = LogisticRegression()

In [51]:
model.fit(X_train, Y_train)

LogisticRegression()

Evaluation of the model

Using Accuracy Score


In [52]:
# Accuracy score on the training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [54]:
print('Accuracy score of the training data : ', training_data_accuracy)

Accuracy score of the training data :  0.9865985576923076


Accuracy on training data at 98.7% is very good 🤓

What's more important is Accuracy Score on the test data, so let's do it

In [55]:
# Accuracy score on the training data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [56]:
print('Accuracy score of the test data : ', test_data_accuracy)

Accuracy score of the test data :  0.9790865384615385


Accuracy Score on the test data at almost 98% is great 🤓

Making a Predictive System

In [57]:
X_new = X_test[0] # storing first row of data to X_new

prediction = model.predict(X_new) # predicting X_new result with the trained model

print(prediction)

if (prediction[0]==0):
  print('The news is real')
else:
  print('The news is fake')

[1]
The news is fake


In [58]:
print(Y_test[0])

1


1 represents as fake is a matching result, so this model has a correct prediction result for X_new